finality-aleph/src/sync/forest/mod.rs

```diff
@@ -0,0 +1,690 @@
+
+pub struct JustificationWithParent<J: Justification> {
+    pub justification: J,
+    pub parent: BlockIdFor<J>,
+}
+
+enum VertexHandle<'a, I: PeerId, J: Justification> {
+    HopelessFork,
+    BelowMinimal,
+    HighestFinalized,
+    Unknown(VacantEntry<'a, BlockIdFor<J>, VertexWithChildren<I, J>>),
+    Candidate(OccupiedEntry<'a, BlockIdFor<J>, VertexWithChildren<I, J>>),
+}
+
+/// Our interest in a block referred to by a vertex, including the information about whom we expect to have the block.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Interest<I: PeerId> {
+    /// We are not interested in this block.
+    Uninterested,
+    /// We would like to have this block.
+    Required(HashSet<I>),
+    /// We would like to have this block and its the highest on its branch.
+    TopRequired(HashSet<I>),
+}
+
+/// What can go wrong when inserting data into the forest.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Error {
+    HeaderMissingParentId,
+    IncorrectParentState,
+    IncorrectVertexState,
+    ParentNotImported,
+}
+
+pub struct VertexWithChildren<I: PeerId, J: Justification> {
+    vertex: Vertex<I, J>,
+    children: HashSet<BlockIdFor<J>>,
+}
+
+impl<I: PeerId, J: Justification> VertexWithChildren<I, J> {
+    fn new() -> Self {
+        Self {
+            vertex: Vertex::new(),
+            children: HashSet::new(),
+        }
+    }
+
+    fn add_child(&mut self, child: BlockIdFor<J>) {
+        self.children.insert(child);
+    }
+}
+
+pub struct Forest<I: PeerId, J: Justification> {
+    vertices: HashMap<BlockIdFor<J>, VertexWithChildren<I, J>>,
+    top_required: HashSet<BlockIdFor<J>>,
+    root_id: BlockIdFor<J>,
+    root_children: HashSet<BlockIdFor<J>>,
+    compost_bin: HashSet<BlockIdFor<J>>,
+}
+
+impl<I: PeerId, J: Justification> Forest<I, J> {
+    pub fn new(highest_justified: BlockIdFor<J>) -> Self {
+        Self {
+            vertices: HashMap::new(),
+            top_required: HashSet::new(),
+            root_id: highest_justified,
+            root_children: HashSet::new(),
+            compost_bin: HashSet::new(),
+        }
+    }
+
+    fn get_mut(&mut self, id: &BlockIdFor<J>) -> VertexHandle<I, J> {
+        use VertexHandle::*;
+        if id == &self.root_id {
+            HighestFinalized
+        } else if id.number() <= self.root_id.number() {
+            BelowMinimal
+        } else if self.compost_bin.contains(id) {
+            HopelessFork
+        } else {
+            match self.vertices.entry(id.clone()) {
+                Entry::Occupied(entry) => Candidate(entry),
+                Entry::Vacant(entry) => Unknown(entry),
+            }
+        }
+    }
+
+    fn prune(&mut self, id: &BlockIdFor<J>) {
+        self.top_required.remove(id);
+        if let Some(VertexWithChildren { children, .. }) = self.vertices.remove(id) {
+            self.compost_bin.insert(id.clone());
+            for child in children {
+                self.prune(&child);
```

Comment:
That's why there are the "long branch" unit tests – the branches should never be longer than two sessions, i.e. 1800 vertices.

Reply:
Ok, fine. So can you perhaps add a comment somewhere that this is the max length and that this guarantee is external to the forest struct... ?

Reply:
Could we maybe add this restriction here? And add vertex state symmetrical to `BelowMinimal`? Something like `OverMaximal`?

Reply:
Hmm, if I added is as an error it would require adding the possibility of an error to a couple methods that don't necessarily need them (and currently don't have them). But I sort of like the idea of having this ensured at this level, so maybe silently dropping whatever is added too high would be reasonable? The silence sucks then though...

So the options I see are:
1. Comment here that this is handled outside – ugly because this is less self-contained, but other than that fine.
2. Add an error – most expressive, but requires lots of error handling where it actually doesn't matter (beyond maybe logging, and even then these will be errors that never turn up).
3. Silently ignore stuff that's too high – ugly because silently failing leads to stuff being extremely hard to debug, but technically we should never even hit these cases and theoretically this is sound, as ignoring these data is well within the model.

After writing this out this way I'm pretty sure 2 is the best option, so I'll start coding it, but if you two have other opinions feel free to complain, it should be easy to rewrite in any of the other options.

Reply:
Silent dropping should not break anything in protocol, right? Whenever we do anything with `hash`, `header`, `justification`, we will check status first, right? So dropping might only make protocol slower, but if we are 900 blocks behind, we are already slow... Only case that would need to be handled is when substrate imports 900 blocks over our top finalized...

Reply:
If you have capacity to write option 2 it is clearly the safest one, so I'm fine with it

Reply:
Added an error then, PTAL.

Reply:
@maciejnems feel free to resolve this last comment once the changes are ok :)